TestVeeamGZipStream/ProcessManager.cs

```diff
@@ -0,0 +1,126 @@
+        private readonly int threadCount = Environment.ProcessorCount;
+
+        private UserThreadPool pool;
+        private ReaderWriterFactory readerWriterFactory;
+
+        #endregion Fields
+
+        #region .ctor
+
+        public ProcessManager(ReaderWriterFactory readerWriterFactory)
+        {
+            this.readerWriterFactory = readerWriterFactory;
+            this.pool = new UserThreadPool(threadCount);
+        }
+
+        #endregion .ctor
+
+        public void Run(CompressionParams settings)
+        {
+            using (var reader = new FileStream(settings.SourceFile,
+            FileMode.Open,
+            FileAccess.Read,
+            FileShare.Read,
+            settings.BlockSize,
+            FileOptions.Asynchronous))
+            {
+                using (var writer = new FileStream(settings.RecoverFileName,
+                    FileMode.OpenOrCreate,
+                    FileAccess.Write,
+                    FileShare.Write,
+                    settings.BlockSize,
+                    FileOptions.Asynchronous))
+                {
+
+                    FileReaderWriter readerWriter = readerWriterFactory.GetFileReaderWriter(reader, writer);
+                    int blockCount;
+                    ///Чтение с конца файла информации о сжатии
+                    if (settings.Mode == Mode.DECOMPRESS)
+                    {
+                        using (MemoryStream compressInfoStream = new MemoryStream())
+                        {
+                            int offset = sizeof(int);
+                            reader.Position = reader.Length - offset;
+                            byte[] countBlocks = new byte[sizeof(int)];
+                            reader.Read(countBlocks, 0, sizeof(int));
+                            blockCount = BitConverter.ToInt32(countBlocks, 0);
+                            //сдвигаем позицию к началу списка размеров блоков
+                            offset += sizeof(int) * blockCount;
+                            reader.Position = reader.Length - offset;
+                            //Проходим каждые 4 байта (размер инта) и считываем значение блока в список
+                            for (int i = 0; i < blockCount; i++)
+                            {
+                                byte[] sizeBlock = new byte[sizeof(int)];
+                                reader.Read(sizeBlock, 0, sizeof(int));
+                                readerWriter.SizeCompressedBlockList.Add(BitConverter.ToInt32(sizeBlock, 0));
+                            }
+
+                            reader.Position = 0;
+                        }
+                        ///Составление задач
+                        foreach (var block in readerWriter.SizeCompressedBlockList)
+                        {
+                            BlockMetadata metadata = new BlockMetadata(block);
+                            Task task = new Task(metadata, settings.Mode.Instruction, readerWriter);
```

Comment:
а здесь вот( это уже будет в Instruction) можно передавать ссылку на метод приватный.
```suggestion
                            Task task = new Task(metadata, () => settings.Mode.Instruction.Apply(), readerWriter);
```
не уверен, что правильно лямбду написал. 
Ну или отставить так же как и есть и будут 2 паблик метода вместо 1 - Apply().